src/sensor/parser.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include <QObject>
+#include "ping-message.h"
+
+// This class digests data and notifies owner when something interesting happens
+class Parser : public QObject
+{
+    Q_OBJECT
+public:
+    Parser() = default;
+    ~Parser() = default;
+
+    enum ParserState {
+        ERROR, // got some unexpected data
+        PARSING, // got some expected data, feed me more
+        NEW_MESSAGE // got a new packet
+    };
+
+    uint32_t parsed; // number of messages/packets successfully parsed
+    uint32_t errors; // number of parse errors
+
+    //TODO move to pure virtual
+    // clear parse state
+    virtual void clearBuffer() { }
+    // asynchronous use, Child classes should signal when something happens ie. 'emit newMessage(Message m)'
+    virtual void parseBuffer(const QByteArray& data) { Q_UNUSED(data) }
+    // synchronous use, Child should return flags indicating incremental parse result/status
+    virtual ParserState parseByte(const char byte) { Q_UNUSED(byte) return ERROR; }
+
+    ping_message rxMessage() { return rxMessage_; }
+
+signals:
+    void newMessage(const ping_message& msg);
+    void parseError();
+
+protected:
+    ping_message rxMessage_;
```

Comment:
Please follow our code style:
https://github.com/bluerobotics/software-guidelines/blob/master/guidelines/style.md#code

Reply:
I was reading some things that prefixing with underscores is not recommended for c/c++, it has some reserved or special meaning.

It will take some time, but I can change it.